src/tools/cephfs/cephfs-shell

```diff
@@ -1100,23 +1100,36 @@ sub-directories, files')
         """
         Disk Usage of a Directory
         """
+        dir_passed = args.dirs
         if args.dirs == '':
-            args.dirs = cephfs.getcwd().decode('utf-8')
-        for dir_ in args.dirs:
-            if args.r:
-                for i in reversed(sorted(set(dirwalk(dir_)))):
-                    i = os.path.normpath(i)
-                    try:
-                        xattr = cephfs.getxattr(to_bytes(i), 'ceph.dir.rbytes')
-                        self.poutput('{:10s} {}'.format(
-                            humansize(int(xattr.decode('utf-8'))), '.' + i))
-                    except libcephfs.Error:
-                        continue
+            dir_passed = cephfs.getcwd().decode('utf-8')
+            if len(dir_passed) > 1:
+                dir_passed = dir_passed.split()
+
+        for dir_ in dir_passed:
+            dir_trav = sorted(set(dirwalk(dir_)))
+            if args.dirs and not dir_trav:
+                dir_trav.append(dir_)
             else:
-                dir_ = os.path.normpath(dir_)
-                self.poutput('{:10s} {}'.format(humansize(int(cephfs.getxattr(
-                    to_bytes(dir_), 'ceph.dir.rbytes').decode('utf-8'))), '.'
-                    + dir_))
+                dir_trav.append('.')
+
+            if args.r:
+                dir_trav = reversed(dir_trav)
```

Comment:
This would print directory in reverse-alphabetic order -

```
CephFS:~/>>> du -r
0B         .
/: no such directory exists
/: no such directory exists
0B         ./d
0B         ./c
0B         ./b/b
0B         ./b
/: no such directory exists
0B         ./a/b
0B         ./a
CephFS:~/>>> 
```

Why do we want that?

Reply:
The unix du does not have that option and I also don't think it is required here. Since, it is present from the initial commit. @batrick can tell us the reason.

Reply:
'r' stands for recursive (not reverse). That is what it was supposed to do in original code.

Reply:
Changes related to this line has introduced a bug.

Contents on FS -
```
/mnt/cephfs1
├── dir1
│   └── dir11
│       └── dir12
│           └── dir13
├── dir2
├── dir3
├── dir4
├── file1
└── slink1 -> file1
```

Copying `du -r` from master -
```
CephFS:~/>>> du -r
0B         .dir4
0B         .dir3
0B         .dir2
0B         .dir1/dir11/dir12/dir13
0B         .dir1/dir11/dir12
0B         .dir1/dir11
0B         .dir1
CephFS:~/>>> 
```

Copying `du -r` from you branch -
```0B         ./dir1
0B         ./dir1/dir11
0B         ./dir1/dir11/dir12
0B         ./dir1/dir11/dir12/dir13
0B         ./dir2
0B         ./dir3
0B         ./dir4
/: no such directory exists
/: no such directory exists
0B         .
```

Reply:
I made those changes to make it similar to unix du, which by default lists recursively.
So, how should cephfs-shell du function?

Reply:
> I made those changes to make it similar to unix du, which by default lists recursively.

Okay. Cool.

> So, how should cephfs-shell du function and why shouldn't it by default list recursively?

Right now du doesn't behave in a recursive fashion, by default.  But you if need to change that maybe you should ask Patrick and/or CephFS team.

Reply:
@varshar16 sorry if I led you down a wrong path in earlier discussion. I now recall we did not want to do recursive listings by default because CephFS clients can be smarter: the recursive statistics ("rstats") available for a directory let us give eventually consistent disk usage for a directory tree. In other words, there's no reason for the client to do a recursive stat listing. So: we do want to have the same behavior with `du(1)` except for this one case because cephfs-shell can be much more efficient.

Reply:
@batrick I've made the suggested change[[1]](https://github.com/ceph/ceph/pull/28560/commits/5e17faf85b8b75b59d6fdd6ad53a9e7931c68324) on my PR since so you left kinda [same review](https://github.com/ceph/ceph/pull/28560#discussion_r302077244) there too few days ago.